pipit/trace.py

```diff
@@ -896,3 +940,156 @@ def detect_pattern(
+        # as well?
+
+        # Amount of time we spend in the label
+        # TODO: can break this down further into CUDA API/kernel launch
+        # time, and other events
+        cpu_time = ann_events.groupby("Name")["time.inc"].sum()
+        print(cpu_time.index)
+
+        ann_kernel_times = pd.Series(
+            [0] * len(cpu_time), index=cpu_time.index, name="time.inc"
+        )
+
+        def _calc_kernel_time(row):
+            # locate the launch event for the kernel
+            # and binary search the start time of the launch
+            # event in annotation events in order to find the
+            # corresponding annotation
+            parent = row["_parent"]
+            # note: parent is always an enter event
+            launch_start = self.events.loc[parent, "Timestamp (ns)"]
+            # we don't use annotation df from above, since we need
+            # leave events also here
+            ann_df = self.events[self.events["type"] == "annotation"]
+            idx = np.searchsorted(ann_df["Timestamp (ns)"], launch_start)
+
+            # update for annotation that launched us
+            # note: searchsorted gives us the leave event (we need the enter event)
+            leave_event = ann_df.iloc[idx]
+            ann_event = ann_df.loc[leave_event["_matching_event"]]
+            if ann_event["Name"] in ann_kernel_times.index:
+                ann_kernel_times.loc[ann_event["Name"]] += row["time.inc"]
+            idx = ann_event["_parent"]
+
+            # update parents of that annotation
+            while (
+                idx in ann_df.index
+                and ann_df.loc[idx, "Name"] in ann_kernel_times.index
+            ):
+                ann_event = ann_df.loc[idx]
+                ann_kernel_times.loc[ann_event["Name"]] += row["time.inc"]
+                idx = ann_event["_parent"]
+
+            # dummy return
+            # TODO: maybe there is a more efficient/cleaner way to do this
+            return 0
+
+        kernels = self.events[
+            (self.events["Event Type"] == "Enter")
+            & self.events["type"].isin(["kernel", "comm"])
+        ]
+        kernels.apply(
+            _calc_kernel_time,
+            axis=1,
+        )
+        ann_time = cpu_time + ann_kernel_times
+
+        # Calculate time in other events
+        # This is sum of exclusive time (total time) - time in
+        # annotations (and kernels launched by those annotations)
+        total_time = pd.concat(
+            [
+                ann_time,
+                pd.Series(
+                    self.events["time.exc"].sum() - ann_time.sum(),
+                    index=["Other"],
+                    name="time.inc",
+                ),
+            ]
+        )
+
+        return total_time
+
+    def filter_by_label(self, label_name):
+        """
+        Filters trace to find kernels
+        that occurred during this time frame
+        and their associated launch events
+        """
```

Comment:
Is this function only applicable on Nsight Systems traces (i.e. traces with GPU kernels)? If so we should mention that in the comment and include an assertion or other check to make sure uses don't try to run it on other traces if it won't work. If it works on other traces maybe we can modify the comment to be more generic.

Reply:
It should still work (the kernels df will just be empty though).

I'll add a test for it, though.